lib/util/stream.js

```diff
@@ -106,11 +106,50 @@ class PartialPipe {
   pipe(out) { return reduce(((x, y) => x.pipe(y)), this.streams[0], this.streams.slice(1)).pipe(out); }
 }
 
+// allWithStreams() is very similar to Promise.all() and returns the same
+// result. The key difference is that if `values` contains a stream or a promise
+// that resolves to a stream, then the fate of the stream will be tied to the
+// fate of each promise within `values`. If any promise is rejected, then the
+// stream will be immediately destroyed.
+const allWithStreams = (values) => {
+  // If/when a promise is rejected, we will need to destroy any streams that we
+  // have already seen (streamsBeforeError), as well as remember to destroy any
+  // streams that we see later (anyError).
+  let anyError = false;
+  const streamsBeforeError = [];
+  const handleStream = (stream) => {
+    if (anyError)
+      stream.destroy();
+    else
+      streamsBeforeError.push(stream);
+  };
+
+  for (const value of values) {
+    Promise.resolve(value)
+      .then((result) => {
+        if (result instanceof PartialPipe) {
+          for (const stream of result.streams) handleStream(stream);
+        } else if (typeof result === 'object' && result != null &&
+          typeof result.pipe === 'function') {
```

Comment:
We need to be able to handle both a regular stream and a `PartialPipe`. It looks like `Submissions.streamForExport()` returns a promise of a `PartialPipe`, while `SubmissionAttachments.streamForExport()` returns a promise of a stream.